### DISCUSSION TITLE
Feature suggestion: pause sync when offline

### DISCUSSION THREAD

**Discussion by obedm503** [maxnowack/signaldb#947] [Ideas]
[I've been playing with the new `SyncManager`](https://github.com/obedm503/trellix-offline) and I have to say I am enjoying it as it puts Signaldb almost on par with other similar libraries.

In that time I noticed some possible areas of improvement. The main ones are network awareness and improvements around remote changes.

It would be nice if the `SyncManager` automatically paused sending `push` requests when the device went offline and resumed via the `online` and `offline` window events. This is technically already possible but it's cumbersome as it involves throwing and catching errors inside the push & pull methods so the sync manager knows the sync failed and retries later.

If you think this is out of scope, the sync manager could expose the `pause` and `pauseAll` methods (opposite of `sync` and `syncAll`) and let the library user implement event listening and cleanup.

Related to this is that it would be nice if the syncManager's `registerRemoteChange` changed how it executes. I believe it should be called for each collection as part of `sync` instead of once in the constructor. This way, it could also receive the `collectionOptions` (just like `push` and `pull`) allowing each collection to connect to an entirely different source or use different configuration options.

Also, the `registerRemoteChange` method should allow returning a cleanup function so we can gracefully shut down the remote data source and avoid memory leaks and other issues.

If `registerRemoteChange` gets called for each collection on `sync` and its cleanup function on `pause`, then going offline and back online would be significantly simpler. It would be a matter of just calling `pauseAll` and `syncAll` on the `offline` and `online` events respectively.

Thanks for creating SignalDB @maxnowack

**Comment by maxnowack** [comment]
I currently handle the offline state exactly like you do in `trellix-offline` (Throwing an error if the device is offline).
I don't think that handling offline/online state in SignalDB is really possible as this check depends a lot on the specific environment. In the browser we have `navigator.onLine` but in native frameworks (React Native, CapacitorJS, etc.) there are different ways how to check the connection state.

I'm not sure yet about changing the behavior of `registerRemoteChange` execution, however I really like the idea of the `pause`/`pauseAll` functionality.
Do you want to hack on it yourself and create a PR or should I implement this? 🙂

**Reply by obedm503** [reply]
Made a draft PR #951. Let me know what you think.
